### PR TITLE
Adding vsphere terraform

### PR DESCRIPTION
**Pull request by michaellevy101**
Upgrading terraform to 0.14
Adding hcl2 library to parse a terraform files
Add vsphere env variables

### CODE REVIEW

File: terraform_files/vsphere/variables-vsphere.tf [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+variable "iso_download_path" {
+  type        = string
+  description = "The ISO path (We have to upload this file to the vsphere)"
+  default = "dummy"
+}
+
+variable "vsphere_cluster" {
+  type        = string
+  description = "vSphere cluster name"
```
Comment (YuviGold) [inline terraform_files/vsphere/variables-vsphere.tf:44]: Explain what's the different usage between the AI cluster name and vSphere cluster name

File: terraform_files/vsphere/variables-vsphere.tf [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+variable "vsphere_username" {
+  type        = string
+  description = "vSphere vcenter server username"
+}
+
+variable "vsphere_password" {
+  type        = string
+  description = "vSphere vcenter server username"
+}
```
Comment (YuviGold) [inline terraform_files/vsphere/variables-vsphere.tf:18]: Rearrange these variables into groups. It's not clear what's mandatory for vSphere login and what's used for other APIs

File: terraform_files/vsphere/variables-vsphere.tf [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  description = "The size of the master's virtual machine's memory, in MB"
+}
+
+variable "master_disk_size_gib" {
+  type = number
+  default = 120
+  description = "The size of the master's disk, in GB"
+}
+
+variable "master_vcpu" {
```
Comment (YuviGold) [inline terraform_files/vsphere/variables-vsphere.tf:77]: Some of these variables are duplicated.
Do we really need to redefine these common variables?
https://github.com/openshift/assisted-test-infra/blob/51d5d1c28822dba48ac56dd4210117d0b606a272/terraform_files/baremetal/variables-libvirt.tf#L131-L135
Could it be a shared variables file?

File: terraform_files/vsphere/variables-vsphere.tf [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  type = number
+  default = 4
+  description = "The total number of virtual processor cores to assign to the master virtual machine."
+}
+
+variable "vsphere_control_plane_cores_per_socket" {
+  type = number
+  default = 1
+  description = <<EOF
+The number of cores per socket in this virtual machine.
```
Comment (YuviGold) [inline terraform_files/vsphere/variables-vsphere.tf:87]: What does a socket represents?

File: terraform_files/vsphere/main.tf [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  required_providers {
+    vsphere = {
+      source = "hashicorp/vsphere"
+      version = "2.0.2"
+    }
+  }
+}
+
+locals {
+  hasISO = var.iso_download_path != "dummy" && var.iso_download_path != null
```
Comment (YuviGold) [inline terraform_files/vsphere/main.tf:11]: In what case would we want a `dummy` download path?

File: terraform_files/vsphere/main.tf [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+data "vsphere_network" "network" {
+  name          = var.vsphere_network
+  datacenter_id = data.vsphere_datacenter.datacenter.id
+}
+
+# Tagging every VirtualMachine, ResourcePool, Folder with the cluster name to easy recognition and add the description:
+# "Created by the test-infra, do not delete manually"
+resource "vsphere_tag_category" "category" {
+  name        = "assisted-test-infra-${var.cluster_name}"
```
Comment (YuviGold) [inline terraform_files/vsphere/main.tf:43]: `cluster_name` already includes `test-infra`

File: terraform_files/vsphere/variables-vsphere.tf (hunk 1 above)
Comment (michaellevy101) [reply terraform_files/vsphere/variables-vsphere.tf:44]: vsphere cluster is a cluster of hosts not applicative cluster.

File: terraform_files/vsphere/variables-vsphere.tf (hunk 3 above)
Comment (michaellevy101) [reply terraform_files/vsphere/variables-vsphere.tf:77]: In the terraform controller we translate between the config and the terraform variables. 
This controller just pass the config as a terraform variables like this:
config = {**self._config.get_all(), **self._cluster_config.get_all(), "cluster_name": self.cluster_name}
and then filter the irrelevant variables. IMO more elegant.

Comment (michaellevy101) [reply terraform_files/vsphere/variables-vsphere.tf:77]: Need to refactor the terraform controller as well but not as part of this epic

File: terraform_files/vsphere/main.tf (hunk 5 above)
Comment (michaellevy101) [reply terraform_files/vsphere/main.tf:11]: never it's just a marker

File: terraform_files/vsphere/main.tf (hunk 6 above)
Comment (michaellevy101) [reply terraform_files/vsphere/main.tf:43]: no, it's not

File: terraform_files/vsphere/main.tf (hunk 5 above)
Comment (michaellevy101) [reply terraform_files/vsphere/main.tf:11]: please read the documentation below

File: terraform_files/vsphere/variables-vsphere.tf (hunk 4 above)
Comment (michaellevy101) [reply terraform_files/vsphere/variables-vsphere.tf:87]: You shouldn't set this variable manually just a default value...
 The number of cores per socket in this virtual machine. The number of vCPUs on the virtual machine will be num_cpus divided by num_cores_per_socket. If specified, the value supplied to num_cpus must be evenly divisible by this value. Default: 1.

Comment (michaellevy101) [reply terraform_files/vsphere/variables-vsphere.tf:87]: did you read the entire section? is from the vsphere official documentation

File: terraform_files/vsphere/variables-vsphere.tf (hunk 2 above)
Comment (YuviGold) [reply terraform_files/vsphere/variables-vsphere.tf:18]: It's still unclear to me what's required for API operations and what's only for login.
In addition, `iso_download_path` shouldn't be under `vSphere variables` category.

File: terraform_files/vsphere/variables-vsphere.tf (hunk 3 above)
Comment (YuviGold) [reply terraform_files/vsphere/variables-vsphere.tf:77]: Please open a ticket

File: terraform_files/vsphere/variables-vsphere.tf (hunk 4 above)
Comment (YuviGold) [reply terraform_files/vsphere/variables-vsphere.tf:87]: I read. I still don't understand what does a socket represents.

File: discovery-infra/test_infra/tools/terraform_utils.py [hunk 7]
```diff
@@ -18,15 +21,35 @@ def __init__(self, working_dir: str):
...
+                                                  input=False, skip_plan=True)
         if return_value != 0:
             message = f"Terraform apply failed with return value {return_value}, output {output} , error {err}"
             logging.error(message)
             raise Exception(message)
 
+    def set_and_apply(self, refresh: bool = True, **kwargs) -> None:
+        defined_variables = self.select_defined_variables(**kwargs)
+        self.change_variables(defined_variables)
+        self.apply(refresh=refresh)
```
Comment (YuviGold) [inline discovery-infra/test_infra/tools/terraform_utils.py:51]: How those modifications are related to VSphere?

**Review by YuviGold — COMMENTED** [review]
Add a README providing a holistic overview about using vsphere provider.
You should direct to the original docs https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs

but also explain how it integrates with our current infra.

File: terraform_files/vsphere/variables-vsphere.tf (hunk 3 above)
Comment (michaellevy101) [reply terraform_files/vsphere/variables-vsphere.tf:77]: https://issues.redhat.com/browse/MGMT-7232?filter=-2

File: discovery-infra/test_infra/tools/terraform_utils.py (hunk 7 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/tools/terraform_utils.py:51]: This code allow us to take the configuration objects, select only the relevant terraform properties and then apply.
It's used by the controller but basically it's a utility method.
I'm trying to splitting out things

**Comment by michaellevy101** [comment]
@YuviGold you asked for a documentation for the vsphere terraform  
It's used internally by the controller you shouldn't use it manually. it's just a code which has a documentation inside it.
The provider name is there so you can just google it and i'm expecting a basic terraform knowledge

File: terraform_files/vsphere/variables-vsphere.tf (hunk 1 above)
Comment (YuviGold) [reply terraform_files/vsphere/variables-vsphere.tf]: Don't explain it to me.. Write it on the description

Comment (michaellevy101) [reply terraform_files/vsphere/variables-vsphere.tf]: I think we shouldn't describe vsphere object types but i did

**Comment by YuviGold** [comment]
> @YuviGold Readme file isn't necessary because vsphere terraform files are used internally by the controller you shouldn't use it manually. it's just a code which has a documentation inside it.
> The provider name is there so you can just google it and i'm expecting a basic terraform knowledge
> IMO we should document manual steps or non obvious code something that requires a second look
> 
> We shouldn't document vsphere objects like cluster, vm, folder and tag nor terraform vsphere provider just our usage which is a trivial (self documenting code) expect the special cases that i documented in the code.
> 
> if you feel that something isn't trivial please add a comment there and will try to document it better

I think it could be nice to have an overview of the integration process between AI and vSphere somewhere.

File: terraform_files/vsphere/variables-vsphere.tf (hunk 4 above)
Comment (michaellevy101) [reply terraform_files/vsphere/variables-vsphere.tf:93]: please open a ticket for vmware:) https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine#num_cores_per_socket

**Comment by michaellevy101** [comment]
https://github.com/openshift/assisted-test-infra/pull/958#issuecomment-879825028
agreed, after the whole process would work i will create a short readme but no need to document the terraform internally.
We should document the controller itself and how to to create a multi-platform test

**Comment by michaellevy101** [comment]
> > @YuviGold Readme file isn't necessary because vsphere terraform files are used internally by the controller you shouldn't use it manually. it's just a code which has a documentation inside it.
> > The provider name is there so you can just google it and i'm expecting a basic terraform knowledge
> > IMO we should document manual steps or non obvious code something that requires a second look
> > We shouldn't document vsphere objects like cluster, vm, folder and tag nor terraform vsphere provider just our usage which is a trivial (self documenting code) expect the special cases that i documented in the code.
> > if you feel that something isn't trivial please add a comment there and will try to document it better
> 
> I think it could be nice to have an overview of the integration process between AI and vSphere somewhere.
agreed, after the whole process would work i will create a short readme but no need to document the terraform internally.
We should document the controller itself and how to to create a multi-platform test

**Review by YuviGold — APPROVED** [review]
LGTM

**Review by YuviGold — APPROVED** [review]

**Pull request merged**